### PR TITLE
Add MMR functionality to elasticsearch retriever

### PR DESCRIPTION
**Pull request by sudranga**
Allows MMR functionality only for the case where we have access to the embedding function. Also allows for users to request for fields from elasticsearch store. These are added to the document metadata.

### CODE REVIEW

**Comment by baskaryan** [comment]
cc @joemcelroy

File: libs/langchain/langchain/vectorstores/elasticsearch.py [hunk 1]
```diff
@@ -603,6 +605,65 @@ def similarity_search(
...
+        elif self.vector_query_field not in fields:
+            fields.append(self.vector_query_field)
+        else:
+            remove_vector_query_field_from_metadata = False
+
+        # Embed the query
+        query_embedding = self.embedding.embed_query(query)
+
+        # Fetch the initial documents
+        got_docs = self._search(query=query, k=fetch_k, fields=fields, **kwargs)
```
Comment (joemcelroy) [inline libs/langchain/langchain/vectorstores/elasticsearch.py:650]: suggestion: pass the query_embedding into the named parameter `query_vector` and adjust the conditional in `_search` to use query_vector and not call the embedding function twice

**Comment by joemcelroy** [comment]
Thanks for contribution! Would be good to add a mention that this is possible in documentation too. Happy to add it if you dont manage to.

**Comment by baskaryan** [comment]
> @joemcelroy I'm unable to see the real reason behind the vercel deployment failure. I get a 404 when i click on the link. I suspect it is the integration test that i added. I followed the instructions at ([Recording HTTP interactions with pytest-vcr](https://github.com/langchain-ai/langchain/blob/master/libs/langchain/tests/README.md#recording-http-interactions-with-pytest-vcr)) in https://github.com/langchain-ai/langchain/blob/master/libs/langchain/tests/README.md . Should i actually record the interactions and push it as part of the pull request?

fixed vercel 👍

File: libs/langchain/langchain/vectorstores/elasticsearch.py (hunk 1 above)
Comment (sudranga) [reply libs/langchain/langchain/vectorstores/elasticsearch.py]: Done

File: libs/langchain/langchain/vectorstores/elasticsearch.py [hunk 2]
```diff
@@ -699,6 +762,14 @@ def _search(
         )
 
         hits = [hit for hit in response["hits"]["hits"]]
```
[the remark this answers is not in the document]
Comment (sudranga) [reply libs/langchain/langchain/vectorstores/elasticsearch.py]: Yes. Makes sense. Eliminated the many loops.

**Review by joemcelroy — APPROVED** [review]
Looks good! thanks for contribution.

**Pull request merged**